source/sigv4_quicksort.c

```diff
@@ -0,0 +1,217 @@
+                             size_t high,
+                             size_t itemSize,
+                             ComparisonFunc_t comparator );
+
+/**
+ * @brief A helper function to partition a subarray using the last element
+ * of the array as the pivot. All items smaller than the pivot end up
+ * at its left while all items greater than end up at its right.
+ *
+ * @param[in] pArray The array to be sorted.
+ * @param[in] low The low index of the array.
+ * @param[in] high The high index of the array.
+ * @param[in] itemSize The amount of memory per entry in the array.
+ * @param[out] comparator The comparison function to determine if one item is less than another.
+ *
+ * @return The index of the pivot
+ */
+static size_t partition( void * pArray,
+                         size_t low,
+                         size_t high,
+                         size_t itemSize,
+                         ComparisonFunc_t comparator );
+
+/*-----------------------------------------------------------*/
+
+static void swap( void * pFirstItem,
+                  void * pSecondItem,
+                  size_t itemSize )
+{
+    uint8_t * pFirstByte = pFirstItem;
+    uint8_t * pSecondByte = pSecondItem;
+
+    if( ( pFirstItem != NULL ) && ( pSecondItem != NULL ) )
+    {
+        /* Swap one byte at a time. */
+        while( itemSize-- )
+        {
+            uint8_t tmp = *pFirstByte;
+            *pFirstByte = *pSecondByte;
+            ++pFirstByte;
+            *pSecondByte = tmp;
+            ++pSecondByte;
+        }
+    }
+}
+
+static void quickSortHelper( void * pArray,
+                             size_t low,
+                             size_t high,
+                             size_t itemSize,
+                             ComparisonFunc_t comparator )
+{
+    size_t stack[ SIGV4_WORST_CASE_SORT_STACK_SIZE ];
+    /* Low and high are first two items on the stack. */
+    size_t top = 0;
+
+    PUSH_STACK( low, stack, top );
+    PUSH_STACK( high, stack, top );
+
+    while( top > 0 )
+    {
+        size_t partitionIndex;
+        size_t lo1, lo2, hi1, hi2;
+        size_t len1, len2;
+        POP_STACK( high, stack, top );
+        POP_STACK( low, stack, top );
+
+        partitionIndex = partition( pArray, low, high, itemSize, comparator );
+
+        len1 = ( ( partitionIndex != 0U ) && ( partitionIndex - 1U > low ) ) ? partitionIndex - 1U - low : 0U;
+        len2 = ( partitionIndex + 1U < high ) ? high - partitionIndex - 1U : 0U;
+
+        if( len1 >= len2 )
+        {
+            lo1 = low;
+            hi1 = partitionIndex - 1U;
+            lo2 = partitionIndex + 1U;
+            hi2 = high;
+        }
+        else
+        {
+            lo1 = partitionIndex + 1U;
+            hi1 = high;
+            lo2 = low;
+            hi2 = partitionIndex - 1U;
+            /* Also swap the lengths so len1 > len2. */
+            len1 ^= len2;
+            len2 ^= len1;
+            len1 ^= len2;
```

Comment:
Oh you actually used my xor swap suggestion. An alternative without the lo/hi variables would be
```c
if( len1 > len2 )
{
    PUSH_STACK( low, stack, top );
    PUSH_STACK( partitionIndex - 1U, stack, top );

    if( len2 > 0U )
    {
        PUSH_STACK( partitionIndex + 1U, stack, top );
        PUSH_STACK( high, stack, top );
    }
}
else
{
    if( len2 > 0U )
    {
        PUSH_STACK( partitionIndex + 1U, stack, top );
        PUSH_STACK( high, stack, top );
    }

    if( len1 > 0U )
    {
        PUSH_STACK( low, stack, top );
        PUSH_STACK( partitionIndex - 1U, stack, top );
    }
}
```
It's up to your preference whether you prefer having the four extra variables and less `if`s, or fewer variables but more `if` statements